helm-q.org

```diff
@@ -476,6 +480,84 @@ Argument ARG: prefix argument."
     (helm :sources (helm-make-source "helm-q" 'helm-q-source)
           :buffer "*helm q*")))
 #+END_SRC
+** test connecting in q-mode
+*** requirement
+We will try to send a command after connecting via ~q~ or ~qcon~ in Emacs, and execute different actions based on the test.
+
+The current behaviour of qcon for a command likes this.
+#+begin_src sh
+$ qcon 208.110.70.170:5000 # a normal successful connection without password.
+208.110.70.170:5000>1+1
+2
+208.110.70.170:5000>\\
+$ qcon 208.110.70.170:5010:admin:pwd # a normal successful connection with password.
+208.110.70.170:5010>2
+2
+208.110.70.170:5010>\\
+$ qcon 208.110.70.170:5010:admin:badpwd # an invalid connection with bad password.
+208.110.70.170:5010>2
+208.110.70.170:5010>\\
+$ qcon 208.110.70.170:5010:baduser:badpwd  # an invalid connection with bad user.
+208.110.70.170:5010>2
+208.110.70.170:5010>\\
+$ qcon 208.110.70.170:5020                 # no process on the port
+208.110.70.170:5020>2
+conn: Connection refused
+$ qcon 208.110.70.171:5000                 # bad host
+208.110.70.171:5000>2
+conn: No route to host
+#+end_src
+So the actions we will do based on a test command are.
+- if after a command and the process is still alive,
+  - if there is a response,
+    we will treat it as a successful connection and store the username and password based on current storage method.
+  - if there is no response,
+    we will treat it as a failed connection and ask for new username and password for it and connect again.
+- if the process is dead,
+  - we will do nothing with it.
+
+*** add a filter function to q-mode
+To achieve the above goal, we will and a new comint output filter function in ~q-shell-mode~ to run the test command.
+#+BEGIN_SRC elisp
+(defun helm-q-mode-hook ()
+  "Run a hook."
+  (add-hook (make-local-variable 'comint-output-filter-functions) 'helm-q-comint-output-filter ))
+(add-hook 'q-shell-mode-hook 'helm-q-mode-hook)
+#+END_SRC
+*** comint output filter to test connecting.
+#+BEGIN_SRC elisp
+(defvar-local helm-q-test-command-sent-p nil "Whether we have sent out the test command.")
+(defvar helm-q-test-command "\"Test connection...\"")
+(defvar helm-q-test-done "\"Test connection...done.\"")
+(defun helm-q-comint-output-filter (message)
```

Comment:
can the connection testing functionality be isolated away from any helm and password management functionality ("update user name and password" ...) ?
maybe factor it out into a `comint-q-connection-test` function returning either true or false, which can be called from `helm-q.el` . thus isolated, without any reference to password management or helm, the definition will be prepared for easy adoption upstream by the `q-mode` project author.

Reply:
Yes

> can the connection testing functionality be isolated away from any helm and password management functionality ("update user name and password" ...) ?
> maybe factor it out into a `comint-q-connection-ok` function returning either true or false, which can be called from `helm-q.el` . thus isolated, without any reference to password management or helm, the definition will be prepared for easy adoption upstream by the `q-mode` project author.

Yes, we will do it so.